File: test/language/templates/cpp/StructTemplateInTemplateTest.cpp
#include "gtest/gtest.h"

#include "templates/struct_template_in_template/StructTemplateInTemplate.h"

#include "zserio/RebindAlloc.h"

namespace templates
{
namespace struct_template_in_template
{

using allocator_type = StructTemplateInTemplate::allocator_type;
using string_type = zserio::string<allocator_type>;

TEST(StructTemplateInTemplateTest, readWrite)
{
    StructTemplateInTemplate structTemplateInTemplate;
    structTemplateInTemplate.setUint32Field(Field_uint32{Compound_uint32{42}});
    structTemplateInTemplate.setStringField(Field_string{Compound_string{string_type{"string"}}});

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    structTemplateInTemplate.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    StructTemplateInTemplate readStructTemplateInTemplate(reader);

    ASSERT_TRUE(structTemplateInTemplate == readStructTemplateInTemplate);
}

} // namespace struct_template_in_template
} // namespace templates

File: test/language/templates/python/StructTemplateInTemplateTest.py
import zserio

import Templates

class StructTemplateInTemplateTest(Templates.TestCase):
    def testReadWrite(self):
        structTemplateInTemplate = self.api.StructTemplateInTemplate(
            self.api.Field_uint32(self.api.Compound_uint32(42)),
            self.api.Field_string(self.api.Compound_string("string"))
        )

        writer = zserio.BitStreamWriter()
        structTemplateInTemplate.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readStructTemplateInTemplate = self.api.StructTemplateInTemplate()
        readStructTemplateInTemplate.read(reader)
        self.assertEqual(structTemplateInTemplate, readStructTemplateInTemplate)
